lib/index.js

```diff
@@ -74,20 +108,39 @@ function OniyiHttpClient(params = {}) {
     );
 
     // augment requestParams to properly handle absolute uris
-    requestParams = parseUri(requestParams);
-
-    // call request directly, if there are no plugins loaded in this client
-    if (!Array.isArray(plugins) || plugins.length < 1) {
-      logger.debug('no plugins loaded, uri: %s', initialParams.uri);
-      return new request.Request(requestParams);
-    }
+    requestOptions = parseUri(requestOptions);
+
+    const ctx = { requestOptions, hookState };
+    // save reference to the original callback
+    const { callback } = requestOptions;
+    // https://github.com/kriskowal/q#using-deferreds
+    const { promise, resolve, reject } = q.defer();
+
+    requestPhaseList.run(ctx, (phaseError) => {
+      if (phaseError) {
+        return _.isFunction(callback)
+          ? callback(phaseError)
```

Comment:
so what happens with the promise in this case? it gets neither resolved nor rejected?

Reply:
hmm right... we can request for callback and promise together. Can we return both?
```
if (phaseError) {
   callback(phaseError);
   reject(phaseError);
   return
}

Reply:
actually, if callback is defined, return both, otherwise just reject with error?

Reply:
can we use sth similar to the `callbackPromise()` that we use in our gravity-repos? So when callback is invoked with error-first, promise is rejected.

maybe we publish callbackPromise as it's own module?

Reply:
I thought about that too. This could work for resolving error, as you explained. 
The problem is that we want to resolve a promise as soon as Requst instance has been created:
```js
const opt.callback = () => {};
resolve(new Request(opt));
```

And resolving callback happens inside of this Request(after the promise has been returned).
Since these are not happening at the same time, I think that we can't use `callbackPromise()`.

Reply:
this error would happen even before the request object is created. and `options.callback` has a totally different purpose than the promise that we return from `makeRequest()`. The callback is to handle an actually received response (the result from the http request being made) while the promise is to be able to async modify request options in pre-flight and still somehow return the request stream to the caller for `makeRequest()`.
So, when any of the plugins in pre-flight fail / produce an error, do we actually need to report the error twice? I think "yes" – because the caller might not be interested in the stream and thus won't handle the promise at all and so wouldn't realise that something went wrong... and his callback is never called too.

Maybe we can temporarily hijack `options.callback` and add a promise rejection handler to also invoke callback with that rejection reason. Additionally prevent `options.callback` from being invoked more than once.

Reply:
> this error would happen even before the request object is created.

phaseList.run() executes after the request object is being created. Which means that error must happen after requestObject gets created. Are we talking about the same thing here?

> options.callback has a totally different purpose than the promise that we return from 
makeRequest()

correct.

> So, when any of the plugins in pre-flight fail / produce an error, do we actually need to report the error twice? I think "yes"

Why do we need to report error twice? user can invoke `makeRequst()` and wait for callback or promise... but not both, right?  I've also added some tests that validate error handling (both by callback and promise)

With the current implementation, if caller needs a stream, he does not provide a callback, and vice versa.

If both are required, this `stream` can be used for listening for "complete"/"error" events, right?